repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DomainPreProcessor.java

```diff
@@ -0,0 +1,164 @@
+import static org.apache.atlas.repository.util.AtlasEntityUtils.mapOf;
+
+public class DomainPreProcessor extends AbstractDomainPreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(DomainPreProcessor.class);
+    private AtlasEntityHeader parentDomain;
+    private EntityGraphMapper entityGraphMapper;
+    private EntityMutationContext context;
+
+    public DomainPreProcessor(AtlasTypeRegistry typeRegistry, EntityGraphRetriever entityRetriever,
+                              AtlasGraph graph, EntityGraphMapper entityGraphMapper) {
+        super(typeRegistry, entityRetriever, graph);
+        this.entityGraphMapper = entityGraphMapper;
+    }
+
+    @Override
+    public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context,
+                                  EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        //Handle name & qualifiedName
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("DomainPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+
+        this.context = context;
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        switch (operation) {
+            case CREATE:
+                processCreateDomain(entity, vertex);
+                break;
+            case UPDATE:
+                processUpdateDomain(entity, vertex);
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void processCreateDomain(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateDomain");
+        String domainName = (String) entity.getAttribute(NAME);
+        String parentDomainQualifiedName = (String) entity.getAttribute(PARENT_DOMAIN_QN);
+
+        domainExists(domainName, parentDomainQualifiedName);
+        entity.setAttribute(QUALIFIED_NAME, createQualifiedName(parentDomainQualifiedName));
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private static String createQualifiedName(String parentDomainQualifiedName) {
+        if (StringUtils.isNotEmpty(parentDomainQualifiedName)) {
+            return parentDomainQualifiedName + "/domain/" + getUUID();
+        } else{
+            return "default/domain" + "/" + getUUID();
+        }
+    }
+
+    private void processUpdateDomain(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateDomain");
+        String vertexQName = vertex.getProperty(QUALIFIED_NAME, String.class);
+        String parentDomainQualifiedName = (String) entity.getAttribute(PARENT_DOMAIN_QN);
```

Comment:
What if while updating, payload does not have parentDomainQualifiedName at all
Handle case, in that case we should read it from vertex